app/controllers/carts_controller.rb

```diff
@@ -0,0 +1,76 @@
+class CartsController < ApplicationController
+  before_action :find_product,
+                :find_item,
+                only: %i(
+                  create
+                  destroy
+                  update_cart
+                  update_rating_item
+                )
+  before_action :logged_in_user, only: %i(index create)
+  before_action :check_quantity, only: :update_cart
+
+  def index
+    @carts = get_all_item_in_cart
+  end
+
+  def create
+    if @item
+      @item["quantity"] += params[:quantity].to_i
+      flash[:success] = t "errors.carts.update"
+    else
+      current_cart << {
+        product: @product,
+        quantity: params[:quantity].to_i,
+        rating: 0
+      }
+      flash[:success] = t "errors.carts.add"
+    end
+    session[:cart] = current_cart
+    redirect_to product_path
+  end
+
+  def destroy
+    if @item
+      current_cart.delete @item
+      flash[:success] = t "errors.carts.remove"
+    end
+    redirect_to carts_path
+  end
+
+  def update_cart
+    if @item
+      @item["quantity"] = params[:quantity].to_i
+      flash[:success] = t "errors.carts.update_success"
+    else
+      flash[:danger] =  t "errors.carts.update_fails"
+    end
+    redirect_to carts_path
+  end
+
+  def update_rating_item
+    if @item
+      @item["rating"] = params[:rating].to_i
+      flash[:success] = t "errors.carts.review_success"
+    else
+      flash[:danger] = t "errors.carts.review_fail"
+    end
+    redirect_to carts_path
+  end
+
+  private
+  def find_product
+    @product = Product.find_by(id: params[:id])
+  end
+
+  def find_item
+    @item = find_product_in_cart @product
+  end
+
+  def check_quantity
+    return unless params["quantity"].to_i > @item["product"]["quantity"]
```

Comment:
code nên dùng if nếu có thể hơn là unless nha em
```suggestion
    return if params["quantity"].to_i <= @item["product"]["quantity"]
```